ports/esp32/machine_pwm.c

```diff
@@ -296,11 +295,18 @@ STATIC int duty_to_ns(machine_pwm_obj_t *self, int duty) {
 #define get_duty_raw(self) ledc_get_duty(self->mode, self->channel)
 
 STATIC uint32_t get_duty_u16(machine_pwm_obj_t *self) {
-    return ledc_get_duty(self->mode, self->channel) << (HIGHEST_PWM_RES + UI_RES_SHIFT - timers[TIMER_IDX(self->mode, self->timer)].duty_resolution);
+    int resolution = timers[TIMER_IDX(self->mode, self->timer)].duty_resolution;
+    int duty = ledc_get_duty(self->mode, self->channel);
+    if (resolution <= UI_RES_16_BIT) {
+        duty <<= HIGHEST_PWM_RES + UI_RES_SHIFT - resolution;
+    } else {
+        duty >>= -(HIGHEST_PWM_RES + UI_RES_SHIFT - resolution);
+    }
+    return duty;
 }
 
 STATIC uint32_t get_duty_u10(machine_pwm_obj_t *self) {
-    return get_duty_u16(self) >> (HIGHEST_PWM_RES - PWRES);
+    return get_duty_u16(self) >> 6;
```

Comment:
The the "6"?  Probably worth a comment, or a #define constant.

Reply:
That#s the difference between 16 bit of duty_u16() and 10 bit of duty(). But I can make it more clear.

Reply:
Yes of course, I didn't see which function this line was in!  It is pretty clear when you see the u10 vs the u16...

Reply:
Changed back to '6' and a comment. That looks more obvious than the macro names.